IMU Pico Code and Classes for Adafruit BNO085

pronav5000 opened the pull request:
I added a new IMU class for the Adafruit BNO085, as well as the Pico code.py for it. The Pico code was tested and confirmed to work. I also updated folder names, file names, and code structure for the Adafruit MMA8451 as well to keep things organized. I did remove the current IMU tests as they were outdated, so I would appreciate some help with that.

christianhbye commented on src/eigsep_sensors/imu/imu.py:
```diff
@@ -0,0 +1,271 @@
+    the I2C bus and sets up the sensor for acceleration measurements.
+    """
+
+    def _parse_line(self, line):
```
Here, you try/except the whole function. It is more explicit to in this case just let an exception be raised if there is one and make downstream users handle it. The risk of doing it the way it's currently implemented is that real errors might be hidden and made hard to debug.

christianhbye commented on scripts/imu/mma8451_read.py:
```diff
@@ -25,17 +16,17 @@
             x, y, z = accelerometer.sensor.acceleration
```
@pronav5000 with the updated code, i think this might fail? It does not seem like your new IMU_MMA8451 class lets you do accelerometer.sensor.acceleration. Did you mean to use read_line like in the other script?

christianhbye commented:
Hi @pronav5000, I am working on tests and I've changed some methods to have the decorator [staticmethod](https://docs.python.org/3/library/functions.html#staticmethod). I have two specific requests

1.  The BN0085 now let's you do cal=True in _request_imu, but the function calling _request_imu (read_iimu) doesn't let you do cal=True. Could you fix this? The leading _ in the name _request_imu indicates that the function shouldn't be called directly by end-users.

2. You have some helper methods: quaternion_to_euler, normalize_vectors, get_pitch_roll_from_gravity etc. Can you figure out (with me and Dominic) which outputs of the IMU we actually care about and the modify the read_imu function to call the appropriate conversion functions? From my other code, I would like to only care about from_imu and have all the other functions be called under the hood if possible.

christianhbye commented on scripts/imu/bno085_code.py:
```diff
@@ -0,0 +1,84 @@
+    time.sleep(1)
+
+    while True:
+        status = imu.calibration_status
```
@pronav5000 Does this line, imu.calibration_status try to do the calibration or does it just read the status? If the latter, it seems like we might be stuck in this while loop forvever in the case where status != 3

pronav5000 replied:
it keeps reading the status, moving around the IMU(rotating it) causes the calibration number to eventually get to 3, ending the loop. I tested the calibration code on the IMU I have wired up and it does not seem to be a problem.

christianhbye commented:
OK, added some tests and made a new file test_imu_math.py where we can add tests for the general math conversion functions. Could you add some documentation for some of the readings? Some keys are obvious (acceleration), but I don't know what stability or steps refer to or what their datatypes are (int, float, str, arrays).

pronav5000 commented:
> OK, added some tests and made a new file test_imu_math.py where we can add tests for the general math conversion functions. Could you add some documentation for some of the readings? Some keys are obvious (acceleration), but I don't know what stability or steps refer to or what their datatypes are (int, float, str, arrays).

I added some documentation in the bno085_code.py. Right now it just prints every property it has, but I feel like the step one will likely not be useful for us, and the stability property could be useful for data analysis later to understand how accurate the readings actually were
